### ISSUE TITLE
Multi run stress test and file descriptors  

### ISSUE THREAD

**Issue by jeroen** [jeroen/curl#150]
A simple benchmark script that downloads all `DESCRIPTION` files from CRAN and writes directly to disk using the internal `file_writer`. Some reservations:

 - The `cran.rstudio.com` mirror has HTTP/2 now so it is faster with HTTP/2 enabled clients.
 - The mirror does server caching so the second time you run it the server is faster.
 - The `cat()` slows things down in rstudio console, better run in terminal.

```r
library(curl)
pkgs <- available.packages("https://cran.rstudio.com/src/contrib")
pool <- new_pool()
dir.create('descriptions', showWarnings = FALSE)
len <- nrow(pkgs)
count <- 0
handles <- lapply(rownames(pkgs), function(x){
  url <- sprintf('https://cran.rstudio.com/web/packages/%s/DESCRIPTION', x)
  file <- sprintf("descriptions/%s.txt", x)
  curl_fetch_multi(url, data = file, fail = function(e){
    cat(e$message, file = stderr())
  }, done = function(){
    count <<- count + 1
    cat(sprintf("\rCompleted %d of %d [%s]                ", count, len, x))
  }, pool = pool)
})

# Much faster the second time due to Coudfront server caching
timings <- system.time(multi_run(pool = pool))
print(timings)
```

For me it downloads 12000 files in approx 40 sec:

```
   user  system elapsed
 22.551  18.796  41.531
```

This makes me somewhat confident that this works on a large scale. 

@hadley @gaborcsardi @wch

**Comment by noamross** [comment]
I did this recently but stopped because of the [CRAN robots.txt file](https://cran.rstudio.com/robots.txt):

```
User-agent: *
Disallow: /web/packages/A3/DESCRIPTION
Disallow: /web/packages/abbyyR/DESCRIPTION
Disallow: /web/packages/abc/DESCRIPTION
Disallow: /web/packages/ABCanalysis/DESCRIPTION
Disallow: /web/packages/abc.data/DESCRIPTION
...
```
(I ended up getting them from the GitHub mirror)

I'm not sure why this is, and I guess specific CRAN mirrors such as RStudio could choose to allow DESCRIPTION scraping. All of the mirrors seem to have it now.  Is it an artifact that should be removed?
